### DISCUSSION TITLE
Illuminate\Foundation\Application useNamespace() method?

### DISCUSSION THREAD

**Discussion by AdamGaskins** [laravel/framework#37833] [Ideas] [answered]
Whenever I start a new application, I like to move my `app/` folder to `src/App` so I can use a structure based on Spatie's [Laravel Beyond Crud](https://spatie.be/products/laravel-beyond-crud). In order to do this I have to extend `\Illuminate\Foundation\Application`:

```php
// src/App/Application.php
namespace App;

class Application extends \Illuminate\Foundation\Application
{
    protected $namespace = 'App\\';
}
```

And then use that in my `bootstrap/app.php` file, while also updating the app path:

```php
$app = (new App\Application(
    realpath(__DIR__.'/../')
))
    ->useAppPath(__DIR__ . '/../src/App');
```

I'm sure there's all kinds of advanced things you can do by extending Application, but it seems a bit overkill just for changing the namespace. I would love to have a `useNamespace()` method on the Application class, which would let me do this:

```php
$app = (new \Illuminate\Foundation\Application(
    realpath(__DIR__.'/../')
))
    ->useAppPath(__DIR__ . '/../src/App')
    ->useNamespace('App\\');
```

I'd be happy to open a PR if I'm not the only one, but thought I'd open a discussion first. Definitely a small thing, but it would make starting a new project that much cleaner :)

**Comment by legecha** [comment, marked as answer]
In Laravel Beyond CRUD, it hsa this paragraph:

> Unfortunately there's one more thing you need to do in order for Laravel to fully support your custom namespaces. By default, Laravel will look at the app/ folder containing all App\ code, and this default is hard coded in the \Illuminate\Foundation\Application class.

But I don't see where this happens (please point that out if I've missed it!). What I *do* see though is this:

https://github.com/laravel/framework/blob/6db59afadca28dfdb2f719e7d79f93885ede17e4/src/Illuminate/Foundation/Application.php#L1383-L1400

It appears the namespace will be grabbed from `composer.json` in the `getNamespace()` method. I realised this after I created the new `Application.php`, added my namespaces to `composer.json` in the psr-4 autoload section and ran `composer dump-autoload`, but actually forgot to update `bootstrap/app.php` to use it!

I had a quick look back at the history of `Application.php` and this code seems to have been in there for over half a decade.

**Reply by AdamGaskins** [reply]
Oh interesting! My initial reaction is that, even though I'm sure the performance impact is negligible, I would like to be able to set the namespace explicitly so it doesn't have to read a file, looping through the namespaces etc. every request. But, I guess this has already been happening in the background without me even noticing so 🤷🏻‍♂️ 

Great find!
